etna/lib/etna/clients/magma/workflows/create_project_from_json_workflow.rb

```diff
@@ -139,20 +149,33 @@ def update_magma_attributes
           end
         end
 
-        def create!
+        def setup_janus_project!
           puts "Creating Janus project."
           create_janus_project
           puts "Done! Adding you as an administrator on the project."
           add_janus_user
           update_janus_permissions
           update_magma_client_token
-          puts "Done! Creating the project in Magma."
+          puts "Done with setting up the project in Janus!"
+        end
+
+        def setup_magma_project!
+          puts "Creating the project in Magma."
           create_magma_project
+          puts "Done! Creating all the models and attributes in Magma."
           create_magma_models
           ensure_magma_tree
+          puts "Done! Adding your new project record."
           create_magma_project_record
           update_magma_attributes
-          puts "All complete! You can visit Janus to refresh your token, then log into any app to manage your data."
+        end
+
+        def create!
+          setup_janus_project!
```

Comment:
I don't think this is idempotent yet...I think Janus will throw an exception if the project name exists, as will Magma. Do you want to add some rescues or try to fetch a project in each system before executing the `setup_` methods? We could also just make it more robust in a follow-on PR, if this one is getting big.

Reply:
@coleshaw I actually added the idempotency to the janus end, it was super simple: https://github.com/mountetna/monoetna/pull/118/files#diff-98c14df942fcf31cc27a6d9b32c06d13ac2882de2e2edcf0022b339160b6368bR97

And magma add project has always been idempotent:  https://github.com/mountetna/monoetna/blob/master/magma/lib/magma/actions/add_project.rb#L8

The big question is the model creation parts.  But I have a bigger question: do we want to remove the creating models from the create project workflow?

SU are not going to include Arjun and folk.  Mostly creating a project will be a me, you, saurab, and gabi operation.  Model creating is an action handled by project leads.  In that case, removing those from this workflow I think is the correct path forward; we scaffold up the project, others add models.

Thoughts?

Reply:
Ah okay, so I missed that! There is another item for idempotency I think: Janus also throws an exception on "DuplicatePermissions", which I ran into this morning when testing ... so that would require a change to that controller.

```
"Duplicate permission on project test17!" (Etna::Error)
```

I'm not so sure that Magma is idempotent? I get a Postgres exception when I try to create a duplicate project on Magma.

```
magma_app_1      | WARN:2020-10-23T17:26:04+00:00 User developer@ucsf.edu calling update_model#action with params {:project_name=>"test18", :actions=>"[{:action_name=>\"add_project\"}]"}
magma_db_1       | 2020-10-23 17:26:04.233 UTC [42] ERROR:  schema "test18" already exists
magma_db_1       | 2020-10-23 17:26:04.233 UTC [42] STATEMENT:  CREATE SCHEMA test18
magma_db_1       | 2020-10-23 17:26:04.234 UTC [42] ERROR:  current transaction is aborted, commands ignored until end of transaction block
magma_db_1       | 2020-10-23 17:26:04.234 UTC [42] STATEMENT:  SAVEPOINT autopoint_1
magma_db_1       | 2020-10-23 17:26:04.234 UTC [42] ERROR:  no such savepoint
magma_db_1       | 2020-10-23 17:26:04.234 UTC [42] STATEMENT:  ROLLBACK TO SAVEPOINT autopoint_1
magma_db_1       | 2020-10-23 17:26:04.235 UTC [42] ERROR:  current transaction is aborted, commands ignored until end of transaction block
magma_db_1       | 2020-10-23 17:26:04.235 UTC [42] STATEMENT:  SAVEPOINT autopoint_1
magma_db_1       | 2020-10-23 17:26:04.235 UTC [42] ERROR:  no such savepoint
magma_db_1       | 2020-10-23 17:26:04.235 UTC [42] STATEMENT:  ROLLBACK TO SAVEPOINT autopoint_1
magma_db_1       | 2020-10-23 17:26:04.235 UTC [42] ERROR:  current transaction is aborted, commands ignored until end of transaction block
magma_db_1       | 2020-10-23 17:26:04.235 UTC [42] STATEMENT:  SAVEPOINT autopoint_1
```

```
/app/lib/etna/client.rb:129:in `status_check!': [{"message":"Unexpected error","source":null,"reason":"PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block\n"}] (Etna::Error)
```

Reply:
Yeah, maybe we split this out into a command we use (setup_janus), and a command that others use (setup_magma) -- I think it's still convenience to be able to set up multiple models at one time, whereas just "add_model" is one at a time, which might be inconvenient.